components/supervisor/pkg/supervisor/services.go

```diff
@@ -38,11 +38,42 @@ type RegisterableRESTService interface {
 	RegisterREST(mux *runtime.ServeMux, grpcEndpoint string) error
 }
 
+type ideReadyService struct {
+	ready bool
+	cond  *sync.Cond
+}
+
+func (service *ideReadyService) wait(ctx context.Context) {
+	service.cond.L.Lock()
+	for !service.ready && ctx.Err() != nil {
```

Comment:
When `Wait` returns, consumers still cannot be sure that the IDE is ready, because `ctx` may have err'd.
Also, the context does not cancel well in here because it's cancelation behaviour depends on the cond for updates.

You could either:
  1. Remove the context from `Wait` if there are no consumers that strictly require the context (don't think that's the case),
  2. or make `Wait(ctx context.Context) (ready bool)`.

Bonus points if you'd wrap the cond wait in a Go routine and have it close a channel. This way you can select `<-ctx.Done()` and `ready= <-ideReady`.